### PR TITLE
Add Radio button and Radio button group

### LINKED ISSUE #23: Epic | Radio Buttons
Overview:
Allow single selection from two or more options.
Doc:
https://modus.trimble.com/components/mobile/radio-buttons/

States:
- Default
- Checked
- Disabled

Stories:

- [ ] #65
- [ ] #66

### LINKED ISSUE #65: Create radio button controls

### LINKED ISSUE #66: Group multiple radio buttons

### PR DESCRIPTION
**Pull request by BAL4-R**
Closes #23, closes #65, closes #66 
Add Radio button and Radio button group
Replaced icons for checkbox to SVG

Mobile: 
image

Windows:
image

### CODE REVIEW

File: Trimble.Modus.Components/Controls/TMRadioButton/TMRadioButton.cs [hunk 1]
```diff
@@ -0,0 +1,244 @@
...
+        private readonly TapGestureRecognizer _tapGesture = new();
+
+        #endregion
+
+        #region Public Properties
+
+        /// <summary>
+        /// Triggered when <see cref="IsSelected"/> changes.
+        /// </summary>
+        public event EventHandler<CheckedChangedEventArgs> SelectedChanged;
```
Comment (rthanga1) [inline Trimble.Modus.Components/Controls/TMRadioButton/TMRadioButton.cs:23]: Is the `CheckedChangedEventArgs` is from the Checkbox? Can we use a different EventArgs for RadioButton? like `SelectionChangedEventArgs`?
And also, please rename the propery name to `SelectionChanged `

File: Trimble.Modus.Components/Controls/TMRadioButton/TMRadioButton.cs [hunk 2]
```diff
@@ -0,0 +1,244 @@
...
+using Trimble.Modus.Components.Enums;
+
+namespace Trimble.Modus.Components
+{
+    public class TMRadioButton : ContentView, IDisposable
+    {
+        #region Fields
+
+        private readonly Label _label;
+        private readonly Image _radio;
```
Comment (rthanga1) [inline Trimble.Modus.Components/Controls/TMRadioButton/TMRadioButton.cs:11]: Can you rename this member name? like '_icon'?

File: DemoApp/DemoApp/Views/TMRadioButtonPage.xaml.cs [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+using Trimble.Modus.Components.Collection;
+using Trimble.Modus.Components.Enums;
+
+namespace DemoApp.Views;
+
+public partial class TMRadioButtonPage : ContentPage
+{
+    private RadioButtonCollection<string> _radioButtons;
+
+    public RadioButtonCollection<string> RadioButtons
```
[the remark this answers is not in the document]
Comment (BAL4-R) [reply DemoApp/DemoApp/Views/TMRadioButtonPage.xaml.cs:14]: We need to override the `InsertItem` function of `ObservableCollection` to handle items being added using `ItemSource.Add(string)` method.

File: Trimble.Modus.Components/Controls/Layouts/WrapLayout.cs [hunk 4]
```diff
@@ -0,0 +1,252 @@
...
+/// <summary>
+/// A layout that arranges views in a wrapping grid.
+/// Ref: https://github.com/mdc-maui/mdc-maui/blob/master/Material.Components.Maui/Components/Layouts/WrapLayout.cs
+/// </summary>
+public partial class WrapLayout : Layout, IItemsLayout
+{
+
+    public static readonly BindableProperty OrientationProperty =
+     BindableProperty.Create(
+         nameof(Orientation),
```
Comment (Sofiya-kumar) [inline Trimble.Modus.Components/Controls/Layouts/WrapLayout.cs:15]: Each file has a different bindable property declaration. Can you please check the CustomButton or TMButton and change this

File: Trimble.Modus.Components/Controls/Layouts/WrapLayout.cs [hunk 5]
```diff
@@ -0,0 +1,252 @@
...
+        set => SetValue(HorizontalSpacingProperty, value);
+    }
+
+    public double VerticalSpacing
+    {
+        get => (double)GetValue(VerticalSpacingProperty);
+        set => SetValue(VerticalSpacingProperty, value);
+    }
+
+    private static void OnSpacingChanged(BindableObject bindable, object oldValue, object newValue)
```
Comment (Sofiya-kumar) [inline Trimble.Modus.Components/Controls/Layouts/WrapLayout.cs:70]: Add regions and follow the regions layer

File: Trimble.Modus.Components/Controls/TMRadioButton/TMRadioButton.cs [hunk 6]
```diff
@@ -0,0 +1,244 @@
...
+        }
+
+        /// <summary>
+        /// Change the enabled state of the button based on IsEnabled property
+        /// </summary>
+        /// <param name="bindable"></param>
+        /// <param name="oldValue"></param>
+        /// <param name="newValue"></param>
+        private static void OnIsEnabledChanged( BindableObject bindable, object oldValue, object newValue)
+        {
```
Comment (Sofiya-kumar) [inline Trimble.Modus.Components/Controls/TMRadioButton/TMRadioButton.cs:152]: Move this methods to private region

**Review by Sofiya-kumar — CHANGES_REQUESTED** [review]

File: Trimble.Modus.Components/Controls/Layouts/WrapLayout.cs (hunk 5 above)
Comment (BAL4-R) [reply Trimble.Modus.Components/Controls/Layouts/WrapLayout.cs:55]: Done

File: Trimble.Modus.Components/Controls/Layouts/WrapLayout.cs (hunk 4 above)
Comment (BAL4-R) [reply Trimble.Modus.Components/Controls/Layouts/WrapLayout.cs]: Done

File: DemoApp/DemoApp/Views/TMRadioButtonPage.xaml [hunk 7]
```diff
@@ -0,0 +1,66 @@
+<?xml version="1.0" encoding="utf-8"?>
+<ContentPage xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
+             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
+             x:Class="DemoApp.Views.TMRadioButtonPage"
+             xmlns:Modus="clr-namespace:Trimble.Modus.Components;assembly=Trimble.Modus.Components"
```
[the remark this answers is not in the document]
Comment (BAL4-R) [reply DemoApp/DemoApp/Views/TMRadioButtonPage.xaml]: Done

File: DemoApp/DemoApp/Views/TMRadioButtonPage.xaml [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+             Title="TMRadioButtonPage">
+  <VerticalStackLayout Spacing="5"
+                       Margin="10">
+    <Label Text="Radio Group"
+           FontAttributes="Bold"
+           FontSize="20" />
+    <Modus:TMRadioButtonGroup x:Name="RadioGroup"
```
[the remark this answers is not in the document]
Comment (BAL4-R) [reply DemoApp/DemoApp/Views/TMRadioButtonPage.xaml]: `x:Name="RadioGroup"` is already in pascal case

File: Trimble.Modus.Components/Controls/TMRadioButton/TMRadioButton.cs (hunk 2 above)
Comment (BAL4-R) [reply Trimble.Modus.Components/Controls/TMRadioButton/TMRadioButton.cs]: Done

File: Trimble.Modus.Components/Controls/TMRadioButton/TMRadioButton.cs (hunk 1 above)
Comment (BAL4-R) [reply Trimble.Modus.Components/Controls/TMRadioButton/TMRadioButton.cs]: `CheckedChangedEventArgs` is the default EventArg used for RadioButton by maui. Hence using the same.
Renamed `SelectedChanged` to `SelectionChanged`.

File: Trimble.Modus.Components/Controls/Layouts/WrapLayout.cs [hunk 9]
```diff
@@ -0,0 +1,252 @@
+using Microsoft.Maui.Layouts;
+using MIView = Microsoft.Maui.IView;
+
+namespace Trimble.Modus.Components.Controls.Layouts;
+
+/// <summary>
+/// A layout that arranges views in a wrapping grid.
+/// Ref: https://github.com/mdc-maui/mdc-maui/blob/master/Material.Components.Maui/Components/Layouts/WrapLayout.cs
+/// </summary>
+public partial class WrapLayout : Layout, IItemsLayout
```
[the remark this answers is not in the document]
Comment (BAL4-R) [reply Trimble.Modus.Components/Controls/Layouts/WrapLayout.cs]: Done

File: Trimble.Modus.Components/Collection/ObservableCollectionExtensions.cs [hunk 10]
```diff
@@ -0,0 +1,27 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Collections.ObjectModel;
+
+namespace Trimble.Modus.Components.Collection
+{
+    public static class ObservableCollectionExtensions
+    {
+        public static void AddRange<T>(this ObservableCollection<T> collection, IEnumerable<T> items)
```
[the remark this answers is not in the document]
Comment (BAL4-R) [reply Trimble.Modus.Components/Collection/ObservableCollectionExtensions.cs]: Removed

File: Trimble.Modus.Components/Controls/TMRadioButton/TMRadioButton.cs (hunk 6 above)
Comment (BAL4-R) [reply Trimble.Modus.Components/Controls/TMRadioButton/TMRadioButton.cs:159]: Done

File: DemoApp/DemoApp/Views/TMRadioButtonPage.xaml (hunk 8 above)
[the remark this answers is not in the document]
Comment (Sofiya-kumar) [reply DemoApp/DemoApp/Views/TMRadioButtonPage.xaml]: RadioGroup in camel case Can you please change it to pascal case "radioGroup"

File: DemoApp/DemoApp/Views/TMRadioButtonPage.xaml.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (rthanga1) [reply DemoApp/DemoApp/Views/TMRadioButtonPage.xaml.cs]: Hope this is resolved now

File: Trimble.Modus.Components/Controls/TMRadioButton/TMRadioButtonGroup.cs [hunk 11]
```diff
@@ -0,0 +1,501 @@
...
+        group.PopulateItems();
+    }
+
+    /// <summary>
+    /// Populate radiobuttons in items source in the group
+    /// </summary>
+    private void PopulateItems()
+    {
+        if (ItemsSource == null) return;
+        foreach (var item in ItemsSource)
```
Comment (rthanga1) [inline Trimble.Modus.Components/Controls/TMRadioButton/TMRadioButtonGroup.cs:338]: Don't we have to remove all the old children first before adding new children?

File: Trimble.Modus.Components/Controls/TMRadioButton/TMRadioButtonGroup.cs [hunk 12]
```diff
@@ -0,0 +1,501 @@
...
+        return -1;
+    }
+
+    /// <summary>
+    /// Loops through all the children of the radio button group
+    /// and sets the IsSelected property of the radio buttons in the group
+    /// </summary>
+    /// <param name="children">List of children's in <see cref="TMRadioButtonGroup"/> </param>
+    /// <param name="index">Index of the selected <see cref="TMRadioButton"/></param>
+    private void LoopChildren(IList<IView> children, int index)
```
Comment (rthanga1) [inline Trimble.Modus.Components/Controls/TMRadioButton/TMRadioButtonGroup.cs:477]: This method, not only loops the children, but also changes it's IsSelectted property. Can you rename this method to `SelectChildrenAtIndex` or something like that?

File: Trimble.Modus.Components/Controls/TMRadioButton/TMRadioButtonGroup.cs [hunk 13]
```diff
@@ -0,0 +1,501 @@
...
+    {
+        foreach (var child in children)
+        {
+            if (child is TMRadioButton radioButton)
+            {
+                radioButton.IsSelected = index == SelectedIndex;
+                index++;
+            }
+            else if (child is Layout childLayout)
```
Comment (rthanga1) [inline Trimble.Modus.Components/Controls/TMRadioButton/TMRadioButtonGroup.cs:486]: Do we need this else if now? as we raise an exception when anything other than `TMRadioButton` is added to the group

File: Trimble.Modus.Components/Controls/TMRadioButton/TMRadioButtonGroup.cs [hunk 14]
```diff
@@ -0,0 +1,501 @@
...
+        SelectedRadioButtonChangedCommand?.Execute(eventArgs);
+        SelectedRadioButtonChanged?.Invoke(this, eventArgs);
+    }
+
+    /// <summary>
+    /// Finds the index of the selected radio button in the group
+    /// </summary>
+    /// <param name="radioButton">Selected Radio button</param>
+    /// <returns>integer index</returns>
+    private int FindIndex(TMRadioButton radioButton)
```
Comment (rthanga1) [inline Trimble.Modus.Components/Controls/TMRadioButton/TMRadioButtonGroup.cs:435]: Do we need this method and the `FindIndexRecursive` now? As we will have only the `TMRadioButton` as the children, can't we use just the `FindIndex(x => x.Text == buttonTitle);`?

File: DemoApp/DemoApp/Views/TMRadioButtonPage.xaml.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (BAL4-R) [reply DemoApp/DemoApp/Views/TMRadioButtonPage.xaml.cs]: Yes

File: Trimble.Modus.Components/Controls/TMRadioButton/TMRadioButtonGroup.cs (hunk 14 above)
Comment (BAL4-R) [reply Trimble.Modus.Components/Controls/TMRadioButton/TMRadioButtonGroup.cs]: Not needed. Changed to `_radioButtons.IndexOf(selectedRadioButton);`

File: Trimble.Modus.Components/Controls/TMRadioButton/TMRadioButtonGroup.cs (hunk 13 above)
Comment (BAL4-R) [reply Trimble.Modus.Components/Controls/TMRadioButton/TMRadioButtonGroup.cs]: Not needed, removed

File: Trimble.Modus.Components/Controls/TMRadioButton/TMRadioButtonGroup.cs (hunk 12 above)
Comment (BAL4-R) [reply Trimble.Modus.Components/Controls/TMRadioButton/TMRadioButtonGroup.cs]: Done

File: Trimble.Modus.Components/Controls/TMRadioButton/TMRadioButtonGroup.cs (hunk 11 above)
Comment (BAL4-R) [reply Trimble.Modus.Components/Controls/TMRadioButton/TMRadioButtonGroup.cs:344]: Yes, done

**Review by rthanga1 — APPROVED** [review]

**Review by Sofiya-kumar — APPROVED** [review]

**Pull request merged**